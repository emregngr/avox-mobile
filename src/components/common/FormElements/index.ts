export * from './CheckboxField'
export * from './TextInputField'
