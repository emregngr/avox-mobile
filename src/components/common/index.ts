export * from './FaqItem'
export * from './FormElements'
export * from './FullScreenLoading'
export * from './Header'
export * from './OnboardingItem'
export * from './ProfileItem'
export * from './RenderDetailTabBar'
export * from './RenderTabBar'
export * from './SafeLayout'
export * from './SearchInput'
export * from './ThemedButton'
export * from './ThemedButtonText'
export * from './ThemedGradientButton'
export * from './ThemedText'
export * from './ToastConfig'
