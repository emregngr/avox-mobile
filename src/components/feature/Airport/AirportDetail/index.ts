export * from './AirportHeader'
export * from './AirportRowItem'
export * from './AirportSectionRow'
export * from './Tab'
