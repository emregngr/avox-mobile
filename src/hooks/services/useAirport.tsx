import { useQuery } from '@tanstack/react-query'
import { useEffect, useMemo, useState } from 'react'

import { getAllAirports } from '@/services/airportService'
import useLocaleStore from '@/store/locale'
import type { Airport } from '@/types/feature/airport'
import { useDebounce } from '@/utils/common/useDebounce'
import { parseFilterRange } from '@/utils/feature/parseFilterRange'

const ITEMS_PER_PAGE = 20

export function useAirport() {
  const { selectedLocale } = useLocaleStore()
  const [searchTerm, setSearchTerm] = useState<string>('')
  const [filters, setFilters] = useState<any>({})
  const [page, setPage] = useState<number>(1)
  const [isSearchLoading, setIsSearchLoading] = useState<boolean>(false)
  const debouncedSearch = useDebounce<string>(searchTerm, 500)

  const {
    data: allAirportsData,
    error,
    isLoading,
  } = useQuery({
    queryFn: () => getAllAirports(selectedLocale),
    queryKey: ['allAirports', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  const applyFilters = (airports: Airport[]): Airport[] => {
    if (!airports) return []

    return airports.filter(airport => {
      if (!airport) return false

      if (filters?.region) {
        const airportRegion = airport?.operations?.region?.toLowerCase()
        const filterRegion = filters?.region?.toLowerCase()

        if (airportRegion !== filterRegion) {
          return false
        }
      }

      if (filters?.airportType) {
        const airportType = airport?.operations?.airportType?.toLowerCase()
        const filterAirportType = filters?.airportType?.toLowerCase()

        if (airportType !== filterAirportType) {
          return false
        }
      }

      const airportServices = airport?.facilities?.services?.map(s => s?.toLowerCase()) || []

      const serviceKeywords: Record<string, string[]> = {
        hasCarRental: ['car rental'],
        hasChildrensPlayArea: ["children's play area"],
        hasDutyFree: ['duty-free shops'],
        hasHealthServices: ['health services'],
        hasHotels: ['hotels'],
        hasLounges: ['lounges'],
        hasPrayerRoom: ['prayer rooms'],
        hasRestaurants: ['restaurants'],
      }

      for (const serviceName in serviceKeywords) {
        if (filters?.[serviceName] === true) {
          const keywords = serviceKeywords?.[serviceName]

          const hasMatchingService = airportServices?.some(serviceString =>
            keywords?.some((keyword: string) => serviceString?.includes?.(keyword)),
          )

          if (!hasMatchingService) {
            return false
          }
        }
      }

      if (filters?.is24Hour === true && !airport?.operations?.is24Hour) return false
      if (filters?.is24Hour === false && airport?.operations?.is24Hour) return false

      if (filters?.hasMetro === true && !airport?.facilities?.hasMetro) return false
      if (filters?.hasMetro === false && airport?.facilities?.hasMetro) return false

      if (filters?.freeWifi === true && !airport?.facilities?.freeWifi) return false
      if (filters?.freeWifi === false && airport?.facilities?.freeWifi) return false

      if (filters?.minGoogleRating && airport?.facilities?.googleMapsRating) {
        if (airport?.facilities?.googleMapsRating < filters?.minGoogleRating) {
          return false
        }
      }

      const rangeChecks: {
        filterKey: keyof typeof filters
        isYear?: boolean
        jsonPath: (a: Airport) => number | undefined | null
      }[] = [
        {
          filterKey: 'foundingYear',
          isYear: true,
          jsonPath: a =>
            a?.airportInfo?.foundingYear
              ? parseInt(String(a?.airportInfo?.foundingYear), 10)
              : undefined,
        },
        { filterKey: 'passengerCapacity', jsonPath: a => a?.infrastructure?.passengerCapacity },
        { filterKey: 'employeeCount', jsonPath: a => a?.airportInfo?.employeeCount },
        { filterKey: 'destinationCount', jsonPath: a => a?.flightOperations?.destinationCount },
        {
          filterKey: 'destinationCountries',
          jsonPath: a => a.flightOperations?.destinationCountries,
        },
        {
          filterKey: 'domesticConnections',
          jsonPath: a => a?.flightOperations?.domesticConnections,
        },
        {
          filterKey: 'internationalConnections',
          jsonPath: a => a?.flightOperations?.internationalConnections,
        },
        { filterKey: 'elevationFt', jsonPath: a => a?.operations?.location?.elevationFt },
        {
          filterKey: 'passengerCapacity',
          jsonPath: a => a?.infrastructure?.passengerCapacity,
        },
        { filterKey: 'baggageCapacity', jsonPath: a => a?.infrastructure?.baggageCapacity },
        { filterKey: 'terminalCount', jsonPath: a => a?.infrastructure?.terminalCount },
        {
          filterKey: 'terminalAreaHectares',
          jsonPath: a => a?.infrastructure?.terminalAreaHectares,
        },
        { filterKey: 'airportAreaHectares', jsonPath: a => a?.infrastructure?.airportAreaHectares },
        { filterKey: 'apronCount', jsonPath: a => a?.infrastructure?.apronCount },
        { filterKey: 'runwayCount', jsonPath: a => a?.infrastructure?.runwayCount },
        { filterKey: 'mainRunwayLengthM', jsonPath: a => a?.infrastructure?.runways?.lengthM },
        { filterKey: 'towerHeightM', jsonPath: a => a?.infrastructure?.towerHeightM },
        {
          filterKey: 'parkingCapacityVehicles',
          jsonPath: a => a?.facilities?.parkingCapacityVehicles,
        },
        { filterKey: 'loungeCount', jsonPath: a => a?.facilities?.loungeCount },
        { filterKey: 'securityQueueTime', jsonPath: a => a?.facilities?.securityQueueTime },
        { filterKey: 'checkinTimeAvg', jsonPath: a => a?.facilities?.checkinTimeAvg },
      ]

      for (const check of rangeChecks) {
        const filterValueStr = filters?.[check.filterKey] as string
        if (filterValueStr) {
          const parsed = parseFilterRange(filterValueStr)
          if (parsed) {
            let itemValue = check?.jsonPath(airport)
            if (itemValue === undefined || itemValue === null || isNaN(Number(itemValue))) {
              return false
            }
            itemValue = Number(itemValue)
            if (itemValue < parsed.min || itemValue > parsed.max) {
              return false
            }
          }
        }
      }
      return true
    })
  }

  const filteredAirports = useMemo(() => {
    if (!allAirportsData) return []
    let filtered = allAirportsData

    if (debouncedSearch.trim()) {
      const searchLower = debouncedSearch.toLowerCase().trim()

      const exactMatches = filtered.filter(
        airport =>
          airport.iataCode?.toLowerCase() === searchLower ||
          airport.icaoCode?.toLowerCase() === searchLower,
      )

      if (exactMatches.length > 0) {
        filtered = exactMatches
      } else {
        filtered = filtered.filter(airport => {
          const name = airport?.name?.toLowerCase() || ''
          const city = airport?.operations?.location?.city?.toLowerCase() || ''
          const country = airport?.operations?.country?.toLowerCase() || ''
          const iata = airport?.iataCode?.toLowerCase() || ''
          const ident = airport?.icaoCode?.toLowerCase() || ''

          return (
            name.includes(searchLower) ||
            city.includes(searchLower) ||
            country.includes(searchLower) ||
            iata.includes(searchLower) ||
            ident.includes(searchLower)
          )
        })
      }
    }

    return applyFilters(filtered)
  }, [allAirportsData, debouncedSearch, filters])

  const paginatedAirports = useMemo(() => {
    if (!filteredAirports) return []
    return filteredAirports.slice(0, page * ITEMS_PER_PAGE)
  }, [filteredAirports, page])

  const hasMore = paginatedAirports.length < filteredAirports.length

  const loadMore = () => {
    if (hasMore && !isLoading) {
      setPage(prev => prev + 1)
    }
  }

  useEffect(() => {
    if (searchTerm !== debouncedSearch) {
      setIsSearchLoading(true)
    } else {
      setTimeout(() => setIsSearchLoading(false), 300)
    }
  }, [searchTerm, debouncedSearch])

  useEffect(() => {
    setPage(1)
  }, [debouncedSearch, filters])

  return {
    error,
    filteredCount: filteredAirports?.length,
    filters,
    hasMore,
    isLoading,
    isSearchLoading,
    loadMore,
    paginatedAirports,
    searchTerm,
    setFilters,
    setSearchTerm,
  }
}
