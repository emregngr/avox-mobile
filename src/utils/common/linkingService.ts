import { Share } from 'react-native'

import { getLocale } from '@/locales/i18next'
import type { AirlineType } from '@/types/feature/airline'
import type { AirportType } from '@/types/feature/airport'
import { Logger } from '@/utils/common/logger'

const BASE_URL = 'https://avox-redirect.vercel.app/'

export const createAirlineLink = (airline: AirlineType): string => {
  const params = new URLSearchParams({
    id: airline.id,
    name: airline.name,
  }).toString()
  return `${BASE_URL}?type=airline-detail&${params}`
}

export const createAirportLink = (airport: AirportType): string => {
  const params = new URLSearchParams({
    id: airport.id,
    name: airport.name,
  }).toString()
  return `${BASE_URL}?type=airport-detail&${params}`
}

export const shareAirline = async (airline: AirlineType): Promise<void> => {
  try {
    const link = createAirlineLink(airline)
    const message = getLocale('shareAirline', {
      airlineName: airline.name,
      iataCode: airline.iataCode,
      link,
    })

    const title = getLocale('shareAirlineTitle', { airlineName: airline.name })

    await Share.share({
      message,
      title,
    })
  } catch (error) {
    Logger.breadcrumb('shareAirlineError', 'error', error as Error)
  }
}

export const shareAirport = async (airport: AirportType): Promise<void> => {
  try {
    const link = createAirportLink(airport)
    const message = getLocale('shareAirport', {
      airportName: airport.name,
      iataCode: airport.iataCode,
      link,
    })

    const title = getLocale('shareAirportTitle', { airportName: airport.name })

    await Share.share({
      message,
      title,
    })
  } catch (error) {
    Logger.breadcrumb('shareAirportError', 'error', error as Error)
  }
}
