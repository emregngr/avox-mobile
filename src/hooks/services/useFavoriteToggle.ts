import { router } from 'expo-router'
import { useCallback, useMemo } from 'react'

import { useAddFavorite, useIsFavorite, useRemoveFavorite } from '@/hooks/services/useFavorite'
import useAuthStore from '@/store/auth'
import type { FavoriteItemType } from '@/types/feature/favorite'
import { Logger } from '@/utils/common/logger'

export function useFavoriteToggle({ id, type }: FavoriteItemType) {
  const { isAuthenticated } = useAuthStore()

  const { isPending: isAddFavoritePending, mutateAsync: addFavoriteMutation } = useAddFavorite()
  const { isPending: isRemoveFavoritePending, mutateAsync: removeFavoriteMutation } =
    useRemoveFavorite()

  const isFavorite = useIsFavorite({ id, type })

  const isPending = useMemo(
    () => isAddFavoritePending || isRemoveFavoritePending,
    [isAddFavoritePending, isRemoveFavoritePending],
  )

  const handleFavoritePress = useCallback(async (): Promise<void> => {
    if (!isAuthenticated) {
      router.replace('/auth')
      return
    }

    if (isPending) return

    try {
      if (isFavorite) {
        await removeFavoriteMutation({ id, type })
      } else {
        await addFavoriteMutation({ id, type })
      }
    } catch (error) {
      Logger.breadcrumb('Failed to toggle favorite', 'error', error as Error)
    }
  }, [
    isAuthenticated,
    isPending,
    isFavorite,
    id,
    type,
    addFavoriteMutation,
    removeFavoriteMutation,
  ])

  return { handleFavoritePress, isFavorite, isPending }
}
