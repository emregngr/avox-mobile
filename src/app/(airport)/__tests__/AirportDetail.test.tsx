import { fireEvent, render } from '@testing-library/react-native'
import { router, useLocalSearchParams } from 'expo-router'
import type { ReactNode } from 'react'
import React from 'react'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import AirportDetail from '@/app/(airport)/airport-detail'
import { useAirportById } from '@/hooks/services/useAirport'
import { useFavoriteToggle } from '@/hooks/services/useFavoriteToggle'
import useLocaleStore from '@/store/locale'
import useThemeStore from '@/store/theme'
import type { AirportType } from '@/types/feature/airport'
import { shareAirport } from '@/utils/common/linkingService'

jest.mock('@/hooks/services/useAirport')

jest.mock('@/hooks/services/useFavoriteToggle')

jest.mock('@/utils/common/linkingService')

jest.mock('@/store/theme')

const mockedUseThemeStore = useThemeStore as jest.MockedFunction<typeof useThemeStore>

jest.mock('@/store/locale')

const mockedUseLocaleStore = useLocaleStore as jest.MockedFunction<typeof useLocaleStore>

jest.mock('@/components/common', () => {
  const { View, Text, TouchableOpacity } = require('react-native')

  return {
    FullScreenLoading: (props: any) => <View {...props} testID="full-screen-loading" />,

    Header: ({
      title,
      backIconOnPress,
      shareIconOnPress,
      rightIconOnPress,
    }: {
      title: string
      backIconOnPress: () => void
      shareIconOnPress: () => void
      rightIconOnPress: () => void
    }) => (
      <>
        <Text testID="header-title">{title}</Text>
        <TouchableOpacity onPress={backIconOnPress} testID="back-button">
          <Text>Back</Text>
        </TouchableOpacity>
        <TouchableOpacity onPress={shareIconOnPress} testID="share-button">
          <Text>Share</Text>
        </TouchableOpacity>
        <TouchableOpacity onPress={rightIconOnPress} testID="favorite-button">
          <Text>Favorite</Text>
        </TouchableOpacity>
      </>
    ),

    SafeLayout: ({ children }: { children: ReactNode }) => (
      <View testID="safe-layout">{children}</View>
    ),

    RenderDetailTabBar: (props: any) => <View {...props} testID="tab-bar" />,
  }
})

jest.mock('@/components/feature', () => {
  const { View } = require('react-native')

  return {
    AdBanner: () => null,
    AirportHeader: (props: any) => <View {...props} testID="airport-header" />,
    GeneralTab: () => null,
    InfrastructureTab: () => null,
    AirportFlightTab: () => null,
    NearbyPlacesTab: () => null,
  }
})

jest.mock('react-native-collapsible-tab-view', () => ({
  Tabs: {
    Container: ({
      children,
      renderHeader,
    }: {
      children: ReactNode
      renderHeader: () => ReactNode
    }) => (
      <>
        {renderHeader()}
        {children}
      </>
    ),

    Tab: ({ children }: { children: ReactNode }) => <>{children}</>,

    ScrollView: ({ children }: { children: ReactNode }) => <>{children}</>,
  },
}))

const mockedUseLocalSearchParams = useLocalSearchParams as jest.MockedFunction<
  typeof useLocalSearchParams
>

const mockedUseAirportById = useAirportById as jest.MockedFunction<typeof useAirportById>
const mockedUseFavoriteToggle = useFavoriteToggle as jest.MockedFunction<typeof useFavoriteToggle>

const mockedAirport: AirportType = {
  id: '1',
  icaoCode: 'AGGH',
  iataCode: 'HIR',
  isoCountry: 'SB',
  isoRegion: 'SB-GU',
  name: 'Honiara International Airport',
  image: '',
  airportInfo: {
    foundingYear: '1942',
    employeeCount: 250,
    website: 'www.honiara-international-airport.com',
    contactInfo: {
      phone: '+677-123-4567',
      email: 'info@honiara-international-airport.com',
    },
    socialMedia: {
      x: 'https://x.com/honiaraairport',
      linkedin: 'https://www.linkedin.com/company/honiara-international-airport',
      instagram: '',
      tiktok: '',
    },
  },
  operations: {
    airportType: 'small_airport',
    region: 'OC',
    country: 'Solomon Islands',
    location: {
      city: 'Honiara',
      address: 'Honiara, Solomon Islands',
      coordinates: {
        latitude: -9.428,
        longitude: 160.054993,
      },
      elevationFt: 28,
    },
    scheduledService: true,
    is24Hour: true,
  },
  infrastructure: {
    passengerCapacity: 0.5,
    baggageCapacity: 80,
    terminalAreaHectares: 0.5,
    airportAreaHectares: 280,
    runways: {
      lengthM: 2200,
      surface: 'Asphalt',
      pcn: '25/F/B/X/T',
      ilsCategory: 'CAT I',
    },
    apronCount: 3,
    towerHeightM: 20,
    fireCategory: 'Cat 5',
    terminalCount: 1,
    runwayCount: 1,
  },
  facilities: {
    services: [
      'check-in counters',
      'small duty-free shop',
      'basic dining options',
      'lost and found',
    ],
    loungeCount: 1,
    securityQueueTime: 5,
    checkinTimeAvg: 20,
    freeWifi: false,
    googleMapsRating: 3.8,
    parkingCapacityVehicles: 200,
    hasMetro: false,
  },
  flightOperations: {
    destinationCount: 4,
    destinationCountries: 4,
    routes: [
      {
        destinationIata: 'HND',
        frequency: '2x daily',
      },
      {
        destinationIata: 'CTS',
        frequency: 'daily',
      },
      {
        destinationIata: 'ITM',
        frequency: '3x weekly',
      },
    ],
    airlines: ['All Nippon Airways', 'Japan Airlines', 'Jetstar Japan'],
    domesticConnections: 1,
    internationalConnections: 3,
  },
  cargo: {
    annualCargoTonnes: 25000,
    terminalCapacityTonnes: 300,
    coldStorage: true,
    dangerousGoods: true,
  },
  nearbyAttractions: [
    {
      attractionId: '1',
      attractionName: 'Honiara Central Market',
      attractionCoordinates: {
        attractionLatitude: -9.4353,
        attractionLongitude: 159.9556,
      },
      description:
        "A vibrant market offering fresh produce, crafts, and local delicacies, perfect for experiencing Solomon Islands' culture.",
      distanceKm: 10,
    },
    {
      attractionId: '2',
      attractionName: 'Bonegi Beach',
      attractionCoordinates: {
        attractionLatitude: -9.3889,
        attractionLongitude: 159.8667,
      },
      description:
        'Famous for WWII shipwrecks, this beach is a hotspot for diving and history enthusiasts.',
      distanceKm: 20,
    },
    {
      attractionId: '3',
      attractionName: 'National Museum of Solomon Islands',
      attractionCoordinates: {
        attractionLatitude: -9.4333,
        attractionLongitude: 159.9578,
      },
      description:
        'Showcases the history and culture of the Solomon Islands with artifacts and traditional art displays.',
      distanceKm: 10,
    },
    {
      attractionId: '4',
      attractionName: 'Point Cruz',
      attractionCoordinates: {
        attractionLatitude: -9.4333,
        attractionLongitude: 159.9544,
      },
      description:
        'A central hub in Honiara with a yacht harbor and nearby cafes, ideal for a relaxing stroll.',
      distanceKm: 11,
    },
    {
      attractionId: '5',
      attractionName: 'Tenaru Falls',
      attractionCoordinates: {
        attractionLatitude: -9.4167,
        attractionLongitude: 160.0667,
      },
      description:
        'A serene waterfall offering hiking and swimming opportunities in a peaceful natural setting.',
      distanceKm: 2,
    },
    {
      attractionId: '6',
      attractionName: 'Botanical Gardens',
      attractionCoordinates: {
        attractionLatitude: -9.4333,
        attractionLongitude: 159.9589,
      },
      description:
        'A tranquil oasis with tropical plants and walking paths, perfect for nature lovers.',
      distanceKm: 10,
    },
    {
      attractionId: '7',
      attractionName: 'Mataniko Falls',
      attractionCoordinates: {
        attractionLatitude: -9.4167,
        attractionLongitude: 159.9667,
      },
      description:
        'A stunning waterfall ideal for hiking and photography, surrounded by lush greenery.',
      distanceKm: 9,
    },
    {
      attractionId: '8',
      attractionName: 'Honiara War Memorial',
      attractionCoordinates: {
        attractionLatitude: -9.4344,
        attractionLongitude: 159.9578,
      },
      description:
        "A tribute to WWII history, offering a reflective space with views of Honiara's coastline.",
      distanceKm: 10,
    },
    {
      attractionId: '9',
      attractionName: 'Holy Cross Cathedral',
      attractionCoordinates: {
        attractionLatitude: -9.435,
        attractionLongitude: 159.957,
      },
      description:
        'A prominent landmark in Honiara, known for its striking architecture and spiritual significance.',
      distanceKm: 10,
    },
    {
      attractionId: '10',
      attractionName: 'Tulagi Island',
      attractionCoordinates: {
        attractionLatitude: -9.1,
        attractionLongitude: 160.15,
      },
      description:
        'A historic island with WWII relics and diving spots, accessible by boat from Honiara.',
      distanceKm: 37,
    },
  ],
  safety: {
    certifications: ['ICAO Annex 17'],
  },
}

const renderWithSafeAreaProvider = (component: ReactNode) =>
  render(
    <SafeAreaProvider
      initialMetrics={{
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
        frame: { x: 0, y: 0, width: 0, height: 0 },
      }}
    >
      {component}
    </SafeAreaProvider>,
  )

beforeEach(() => {
  mockedUseThemeStore.mockReturnValue({ selectedTheme: 'light' })
  mockedUseLocaleStore.mockReturnValue({ selectedLocale: 'en' })

  mockedUseAirportById.mockReturnValue({ data: null } as any)

  mockedUseFavoriteToggle.mockReturnValue({
    handleFavoritePress: jest.fn(),
    isFavorite: false,
    isPending: false,
  })
})

describe('AirportDetail Screen', () => {
  it('should display loading screen while data is loading', () => {
    mockedUseLocalSearchParams.mockReturnValue({ airportId: '1' })
    mockedUseAirportById.mockReturnValue({ data: null } as any)
    const { getByTestId } = renderWithSafeAreaProvider(<AirportDetail />)
    expect(getByTestId('full-screen-loading')).toBeTruthy()
  })

  it('should display airport details when data is loaded', () => {
    mockedUseLocalSearchParams.mockReturnValue({ airportId: '1' })
    mockedUseAirportById.mockReturnValue({ data: mockedAirport } as any)
    const { getByTestId, queryByTestId } = renderWithSafeAreaProvider(<AirportDetail />)
    expect(getByTestId('airport-header')).toBeTruthy()
    expect(queryByTestId('full-screen-loading')).toBeNull()
  })

  it('should navigate back when the back button is pressed', () => {
    mockedUseLocalSearchParams.mockReturnValue({ airport: JSON.stringify(mockedAirport) })
    const { getByTestId } = renderWithSafeAreaProvider(<AirportDetail />)
    const backButton = getByTestId('back-button')
    fireEvent.press(backButton)
    expect(router.back).toHaveBeenCalledTimes(1)
  })

  it('should call the favorite function when the favorite button is pressed', () => {
    const handleFavoritePress = jest.fn()
    mockedUseFavoriteToggle.mockReturnValue({
      handleFavoritePress,
      isFavorite: false,
      isPending: false,
    })
    mockedUseLocalSearchParams.mockReturnValue({ airport: JSON.stringify(mockedAirport) })
    const { getByTestId } = renderWithSafeAreaProvider(<AirportDetail />)
    fireEvent.press(getByTestId('favorite-button'))
    expect(handleFavoritePress).toHaveBeenCalledTimes(1)
  })

  it('should call the share function when the share button is pressed', async () => {
    mockedUseLocalSearchParams.mockReturnValue({ airport: JSON.stringify(mockedAirport) })
    const { getByTestId } = renderWithSafeAreaProvider(<AirportDetail />)
    fireEvent.press(getByTestId('share-button'))
    expect(shareAirport).toHaveBeenCalledWith(mockedAirport)
  })
})

describe('AirportDetail Screen Snapshot', () => {
  it('should render the AirportDetail Screen successfully', () => {
    mockedUseLocalSearchParams.mockReturnValue({ airport: JSON.stringify(mockedAirport) })

    const { toJSON } = render(<AirportDetail />)

    expect(toJSON()).toMatchSnapshot()
  })
})
