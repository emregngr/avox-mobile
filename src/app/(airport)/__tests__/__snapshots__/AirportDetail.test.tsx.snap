// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AirportDetail Screen Snapshot should render the AirportDetail Screen successfully 1`] = `
<View
  testID="safe-layout"
>
  <Text
    testID="header-title"
  >
    Honiara International Airport
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="back-button"
  >
    <Text>
      Back
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="share-button"
  >
    <Text>
      Share
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="favorite-button"
  >
    <Text>
      Favorite
    </Text>
  </View>
  <RCTScrollView
    contentContainerClassName="flex-1"
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        airportData={
          {
            "airportInfo": {
              "contactInfo": {
                "email": "info@honiara-international-airport.com",
                "phone": "+677-123-4567",
              },
              "employeeCount": 250,
              "foundingYear": "1942",
              "socialMedia": {
                "instagram": "",
                "linkedin": "https://www.linkedin.com/company/honiara-international-airport",
                "tiktok": "",
                "x": "https://x.com/honiaraairport",
              },
              "website": "www.honiara-international-airport.com",
            },
            "cargo": {
              "annualCargoTonnes": 25000,
              "coldStorage": true,
              "dangerousGoods": true,
              "terminalCapacityTonnes": 300,
            },
            "facilities": {
              "checkinTimeAvg": 20,
              "freeWifi": false,
              "googleMapsRating": 3.8,
              "hasMetro": false,
              "loungeCount": 1,
              "parkingCapacityVehicles": 200,
              "securityQueueTime": 5,
              "services": [
                "check-in counters",
                "small duty-free shop",
                "basic dining options",
                "lost and found",
              ],
            },
            "flightOperations": {
              "airlines": [
                "All Nippon Airways",
                "Japan Airlines",
                "Jetstar Japan",
              ],
              "destinationCount": 4,
              "destinationCountries": 4,
              "domesticConnections": 1,
              "internationalConnections": 3,
              "routes": [
                {
                  "destinationIata": "HND",
                  "frequency": "2x daily",
                },
                {
                  "destinationIata": "CTS",
                  "frequency": "daily",
                },
                {
                  "destinationIata": "ITM",
                  "frequency": "3x weekly",
                },
              ],
            },
            "iataCode": "HIR",
            "icaoCode": "AGGH",
            "id": "1",
            "image": "https://res.cloudinary.com/dzkssh0tp/image/upload/c_auto,w_800,q_auto,f_auto/v1758392988/HIR_mnss2t.jpg",
            "infrastructure": {
              "airportAreaHectares": 280,
              "apronCount": 3,
              "baggageCapacity": 80,
              "fireCategory": "Cat 5",
              "passengerCapacity": 0.5,
              "runwayCount": 1,
              "runways": {
                "ilsCategory": "CAT I",
                "lengthM": 2200,
                "pcn": "25/F/B/X/T",
                "surface": "Asphalt",
              },
              "terminalAreaHectares": 0.5,
              "terminalCount": 1,
              "towerHeightM": 20,
            },
            "isoCountry": "SB",
            "isoRegion": "SB-GU",
            "name": "Honiara International Airport",
            "nearbyAttractions": [
              {
                "attractionCoordinates": {
                  "attractionLatitude": -9.4353,
                  "attractionLongitude": 159.9556,
                },
                "attractionId": "1",
                "attractionName": "Honiara Central Market",
                "description": "A vibrant market offering fresh produce, crafts, and local delicacies, perfect for experiencing Solomon Islands' culture.",
                "distanceKm": 10,
              },
              {
                "attractionCoordinates": {
                  "attractionLatitude": -9.3889,
                  "attractionLongitude": 159.8667,
                },
                "attractionId": "2",
                "attractionName": "Bonegi Beach",
                "description": "Famous for WWII shipwrecks, this beach is a hotspot for diving and history enthusiasts.",
                "distanceKm": 20,
              },
              {
                "attractionCoordinates": {
                  "attractionLatitude": -9.4333,
                  "attractionLongitude": 159.9578,
                },
                "attractionId": "3",
                "attractionName": "National Museum of Solomon Islands",
                "description": "Showcases the history and culture of the Solomon Islands with artifacts and traditional art displays.",
                "distanceKm": 10,
              },
              {
                "attractionCoordinates": {
                  "attractionLatitude": -9.4333,
                  "attractionLongitude": 159.9544,
                },
                "attractionId": "4",
                "attractionName": "Point Cruz",
                "description": "A central hub in Honiara with a yacht harbor and nearby cafes, ideal for a relaxing stroll.",
                "distanceKm": 11,
              },
              {
                "attractionCoordinates": {
                  "attractionLatitude": -9.4167,
                  "attractionLongitude": 160.0667,
                },
                "attractionId": "5",
                "attractionName": "Tenaru Falls",
                "description": "A serene waterfall offering hiking and swimming opportunities in a peaceful natural setting.",
                "distanceKm": 2,
              },
              {
                "attractionCoordinates": {
                  "attractionLatitude": -9.4333,
                  "attractionLongitude": 159.9589,
                },
                "attractionId": "6",
                "attractionName": "Botanical Gardens",
                "description": "A tranquil oasis with tropical plants and walking paths, perfect for nature lovers.",
                "distanceKm": 10,
              },
              {
                "attractionCoordinates": {
                  "attractionLatitude": -9.4167,
                  "attractionLongitude": 159.9667,
                },
                "attractionId": "7",
                "attractionName": "Mataniko Falls",
                "description": "A stunning waterfall ideal for hiking and photography, surrounded by lush greenery.",
                "distanceKm": 9,
              },
              {
                "attractionCoordinates": {
                  "attractionLatitude": -9.4344,
                  "attractionLongitude": 159.9578,
                },
                "attractionId": "8",
                "attractionName": "Honiara War Memorial",
                "description": "A tribute to WWII history, offering a reflective space with views of Honiara's coastline.",
                "distanceKm": 10,
              },
              {
                "attractionCoordinates": {
                  "attractionLatitude": -9.435,
                  "attractionLongitude": 159.957,
                },
                "attractionId": "9",
                "attractionName": "Holy Cross Cathedral",
                "description": "A prominent landmark in Honiara, known for its striking architecture and spiritual significance.",
                "distanceKm": 10,
              },
              {
                "attractionCoordinates": {
                  "attractionLatitude": -9.1,
                  "attractionLongitude": 160.15,
                },
                "attractionId": "10",
                "attractionName": "Tulagi Island",
                "description": "A historic island with WWII relics and diving spots, accessible by boat from Honiara.",
                "distanceKm": 37,
              },
            ],
            "operations": {
              "airportType": "small_airport",
              "country": "Solomon Islands",
              "is24Hour": true,
              "location": {
                "address": "Honiara, Solomon Islands",
                "city": "Honiara",
                "coordinates": {
                  "latitude": -9.428,
                  "longitude": 160.054993,
                },
                "elevationFt": 28,
              },
              "region": "OC",
              "scheduledService": true,
            },
            "safety": {
              "certifications": [
                "ICAO Annex 17",
              ],
            },
          }
        }
        testID="airport-header"
      />
    </View>
  </RCTScrollView>
</View>
`;
